### ISSUE TITLE
Alt+Q reflow: behaviour change: text is moved on the same line as the opening brace

### ISSUE THREAD

**Issue by jsmaniac** [racket/gui#47]
With Racket 6.6.0.5 (20160914-62f5b2c):

```
#lang scribble/manual
@defproc[(+ [x integer?] [y integer?]) integer?]{
 AAA BBB CCC DDD EEE FFF GGG HHH III JJJ KKK LLL.}
```

Pressing <kbd>Alt</kbd>+<kbd>Q</kbd> with the cursor somewhere in `AAA …` moves the `AAA …` on the same line as the opening brace:

```
#lang scribble/manual
@defproc[(+ [x integer?] [y integer?]) integer?]{ AAA BBB CCC DDD EEE FFF GGG
 HHH III JJJ KKK LLL.}
```

I think this was not the case before, and I personally preferred the old behaviour, because when the description for a `@defproc` is a large paragraph and the datum+contracts themselves span several lines, having the first line of description on the same line as some of the contracts doesn't make things very clear.

**Comment by rfindler** [comment]
How do you think the code decide where to break the line here?

If this wasn't `@defproc`, but was `@emph{}` then probably we wouldn't want to break, right?

If you have ideas but aren't sure and want to play around, the code is here:

https://github.com/racket/gui/blob/master/gui-lib/scribble/private/indentation.rkt#L281-L325

The painful part of this code is deciding what the the "figure out how much to wrap" algorithm should be (that's where (I think) all the bugs you've found so far is). Currently the "where do I insert new lines" part is straightforward (in comparison). I think you might need to restart drracket to try out changes, but you should be able to stick a few printfs in there to get a sense of what's going on and then experiment with ideas.

Or just make suggestions, I'll try them out (eventually).

**Comment by jsmaniac** [comment]
@rfindler I meant that if a line ends with an opening brace `{`, then it should keep that line break. So:

```
#lang scribble/manual
@defproc[(+ [x integer?] [y integer?]) integer?]{
 AAA BBB CCC DDD EEE FFF GGG HHH III JJJ KKK LLL. AAA BBB CCC DDD EEE FFF GGG HHH III JJJ KKK LLL.}
```

becomes:

```
#lang scribble/manual
@defproc[(+ [x integer?] [y integer?]) integer?]{
 AAA BBB CCC DDD EEE FFF GGG HHH III JJJ KKK LLL. AAA BBB CCC DDD EEE FFF GGG
 HHH III JJJ KKK LLL.}
```

And, following the same rule,

```
#lang scribble/manual
@emph{
 AAA BBB CCC DDD EEE FFF GGG HHH III JJJ KKK LLL. AAA BBB CCC DDD EEE FFF GGG HHH III JJJ KKK LLL.}
```

becomes:

```
#lang scribble/manual
@emph{
 AAA BBB CCC DDD EEE FFF GGG HHH III JJJ KKK LLL. AAA BBB CCC DDD EEE FFF GGG
 HHH III JJJ KKK LLL.}
```

I think that was the reflow behaviour some time ago, as all my documentation is written like this, and I don't remember doing anything special to force the newline after the opening brace (e.g. inserting a blank line, Alt+Q, then removing the blank line).

I'll probably try to fix this at some point, but won't have time until the end of the year (things should settle down work-wise after that). I had a look at the indentation code the other day, and it did seem pretty complex, so I'm sure the "figure out how much to wrap" code is on par with that :) .

Thanks a lot for fixing all those former issues, by the way, I really appreciate it!

**Comment by rfindler** [comment]
The behavior before was unclear and the code is new now, so lets just focus on going forward.

As for that rule, I like the idea in principle, but what happens when you have an `@emph{` that just coincidentally ends up with its `{` as the last character on the line? Then it falls into the other case and things get wonky. 

How do we avoid that problem?

**Comment by jsmaniac** [comment]
@rfindler Yes, that's an issue indeed. I thought that the first word after the brace was moved with it, so that that this:

```
#lang scribble/manual
@;                                                             Limit here ----->|
abc def abc def abc def abc def abc def abc def abc def abc def abc def a @emph{def abc def abc def abc def.}
```

would become this

```
#lang scribble/manual
@;                                                             Limit here ----->|
abc def abc def abc def abc def abc def abc def abc def abc def abc def a
@emph{def abc def abc def abc def.}
```

which would mean that the newline after the brace would not be inserted in the first place, unless the user does so manually, but it's not the case, and the result instead is

```
#lang scribble/manual
@;                                                             Limit here ----->|
abc def abc def abc def abc def abc def abc def abc def abc def abc def a @emph{
 def abc def abc def abc def.}
```

which, I would argue, is better.

------

We could also use a whitelist (like DrRacket has for indentation), but I don't like the idea.


----

I noted that if a line starts with an at-expression, it is handled differently (e.g. the following doesn't reflow to put the `@emph` on the same line as the rest), so I thought that the special behaviour could be "when the `@…` expression started at the beginning of a line **and** the opening brace is the last character on its line (although the `@` and the `{` can be on different lines), then keep the line breaks before the `@` and after the `{`".

```
#lang scribble/manual
abc def
@emph{ghi}
```

 This does require a bit of experimentation, as you said, so I'll try to give it a shot some time later (January, probably, I'm finishing my Ph.D right now so things are a bit crazy).

**Comment by rfindler** [comment]
I like the idea at the end there (if `@` is in the first column and the last column is a `{` then don't bring anything up onto that line). At least I can't find any problems with it by just thinking :)
